languages/scala/runnables.scm

```diff
@@ -39,3 +39,15 @@
     )
     (#set! tag scala-main)
 )
+
+(
+    (
+        (class_definition
+            extend: (extends_clause
+                type: (type_identifier) @run
+            )
+        ) @_scala_test_class_end
+        (#match? @run "^(AnyWordSpec|WordSpec|AnyFunSpec|FunSpec|AnyFunSuite|FunSuite|AnyFlatSpec|FlatSpec|FeatureSpec|AnyFeatureSpec|AnyPropSpec|PropSpec|AnyFreeSpec|FreeSpec)$")
```

Comment:
> Is ZED using code lenses at all? 
I don't know.

> This way seems it might not work in all the adge cases, would be better to leave resolving test classes to Metals as it has the knowledge about what test suites are available.
I don't know how to do that for zed, I took inspiration from the runnables provided by zed:
[https://github.com/zed-industries/zed/blob/main/crates/languages/src/go/runnables.scm](https://github.com/zed-industries/zed/blob/main/crates/languages/src/go/runnables.scm)
 
> People might name something FreeSpec which might not actually be a test
Agreed, it's actually true for the existing queries for App in the existing runnables.scm:
<img width="1509" alt="Screenshot 2024-10-28 at 18 16 15" src="https://github.com/user-attachments/assets/15cb2acc-f11d-4703-86b4-e6682c836e5f">

I tried messing around and also matching an import for `org.scalatest.wordspec.AnyWordSpec` without success.

Reply:
I can't find a mention of [code lenses](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeLens) in Zed docs, so this might no be possible to use. That's the standard way of adding such additional information.

Is it possible to invoke something before showing the runnables or that is purely syntactic? Maybe it would be possible to query code lense and then show them as runnables?

Although looking at what you linked this seems standard to Zed. The go runnables also depend on regex and discovering things named correctly. That seems like a pretty bad idea overall, I wonder what is their explanation for doing that :thinking: 

I guess until code lenses are ready we could go with this workaround :thinking:

Reply:
Though this does depend on Bloop available. Are we able to have this turned on conditionally? So that users can define the task themselves?

Reply:
> Is it possible to invoke something before showing the runnables or that is purely syntactic? Maybe it would be possible to query code lense and then show them as runnables?

I don't think it's possible, you just provide a `runnables.scm` and if a task is available with the same tag the run button is displayed. 

> Though this does depend on Bloop available. Are we able to have this turned on conditionally? So that users can define the task themselves?

The code in the PR does not depend on bloop at all, the task and the script have been provided for testing purposes. 

By default, (it's the same for the main runner) no run icon is displayed, in order to have a run icon displayed, as a zed user, you must add a task with the same tag as the one in the `runnables.scm`. In this particular case, `scala-test`. 

IMHO, the plugin should not decide how the tests are to be run, but it would be a good idea to provide an example in the documentation. It took me some time to understand how to get the Run Icon for tests and main methods...